src/structs/model.rs

```diff
@@ -163,13 +163,13 @@ pub struct ModelPerColor {
 
 #[derive(DefaultBoxed)]
 struct Counts7x7 {
-    exponent_counts: [[Branch; MAX_EXPONENT]; MAX_EXPONENT],
+    exponent_counts: [[Branch; MAX_EXPONENT]; NUMERIC_LENGTH_MAX],
```

Comment:
Why it is needed? This index is defined by the range of bit width of `calc_coefficient_context_7x7_aavg_block` results, and they are maximum 11 bits wide, not 12. Huffman table of 8-bit JPEG has only bit widths up to 11, wider coefficients cannot appear in JPEG file.

Reply:
11 bits wide, but need 12 array slots (one for 0). Not a problem for the exponent array since the 12th slot is implied.

Reply:
Oh, sorry, 11 bits are for DC coef only, these 7x7 AC coefs are at most 10 bits wide, see table F.2 of the standard. And prediction does not extend dynamic range. So still MAX_EXPONENT should be enough.

Reply:
What was the test values that give OOB?

Reply:
Heh, it is even more interesting. In JPEG DC difference is coded, and it can be of -2047 to +2047 (see table F.1). However, Lepton is working with actual (but shifted) DC, that is shorter: actual DC of the block should be in [0,2040] range, but shift of each pixel by -128 following F.1.1.3 makes its range [-1024,1016]. The range wider breaks DC prediction function `adv_predict_or_unpredict_dc`, so it cannot be meaningfully coded by C++ Lepton too.

Reply:
The encoder only restricts the AC to 11 bits not 10, so theoretically the C++ encoder can generate it.

Reply:
> The encoder only restricts the AC to 11 bits not 10, so theoretically the C++ encoder can generate it.

I cannot see how. `calc_coefficient_context_7x7_aavg_block` takes absolute values of coefs, that are coming from JPEG in [0, 1023] range, and does not widen it on any way.

Reply:
Unfortunately there's nothing stopping a nonconforming JPEG from having ACs up to 2047 and the C"" version will try to encode them.

Reply:
If we are talking about nonconforming JPEGs, then in principle Huffman table description allows coefficients up to 16 bit wide. It breaks everything in Lepton with overflows almost everywhere. Should we support them also?

Reply:
It will be a major effort as maximal dequantized coefficient will not fit into i32: 65535*65535 > 2^31.

Reply:
Also "insane" JPEG can have, for example, strictly increasing DC coefs, providing every time difference of +2047 for the next block. It is even conforming to the standard, but will break in `adv_predict_or_unpredict_dc` (see the example above). I'm not sure that we should support all this crazy variants.

Reply:
I agree it sucks to have to handle these corner cases, but if the C++ allows the file (it only disallows abs(coefficients) > 2047), then we still have to be able to decompress it successfully. As a compressor, we have to be careful about not breaking existing compressed files. 

Doesn't mean that we can't overflow, just that the overflow has to be handled in the same way as not to break decompression.

Reply:
You're right, some comments then will be nice here in the code.